components/gitlab/new-milestone.js

```diff
@@ -0,0 +1,91 @@
+const gitlab = require("https://github.com/PipedreamHQ/pipedream/components/gitlab/gitlab.app.js");
+
+module.exports = {
+  name: "New Milestone (Instant)",
+  description: "Triggers when a new milestone is created in a project",
+  version: "0.0.1",
+  dedupe: "unique",
+  props: {
+    gitlab,
+    projectId: { propDefinition: [gitlab, "projectId"] },
+    timer: {
+      type: "$.interface.timer",
+      default: {
+        intervalSeconds: 60,
+      },
+    },
+    http: "$.interface.http",
+    db: "$.service.db",
+  },
+  hooks: {
+    async activate() {
+      const opts = {
+        projectId: this.projectId,
+        pageSize: 1,
+      };
+      const milestones = await this.collectMilestones(opts);
+      if (milestones.length > 0) {
+        const lastProcessedMilestoneId = milestones[0].id
+        this.db.set("lastProcessedMilestoneId", lastProcessedMilestoneId);
+        console.log(
+          `Polling Gitlab milestones created after ID ${lastProcessedMilestoneId}`
+        );
+      }
+    },
+  },
+  methods: {
+    generateMeta(data) {
+      const {
+        id,
+        created_at,
+        title,
+      } = data;
+      const summary = `New milestone created: ${title}`;
+      const ts = +new Date(created_at);
+      return {
+        id,
+        summary,
+        ts,
+      };
+    },
+    async collectMilestones(opts) {
+      const milestones = this.gitlab.getMilestones(opts);
+      const milestonesCollected = [];
+      for await (const milestone of milestones) {
+        milestonesCollected.push(milestone);
+      }
+      return milestonesCollected;
+    },
+  },
+  async run() {
+    // We use the ID of the last processed milestone so that we
+    // don't emit events for them (i.e. we only want to emit events
+    // for new milestones).
+    let lastProcessedMilestoneId = this.db.get("lastProcessedMilestoneId");
+    const opts = {
+      projectId: this.projectId,
+      lastProcessedMilestoneId,
+    };
+    const unprocessedMilestones = await this.collectMilestones(opts);
+
+    if (unprocessedMilestones.length > 0) {
```

Comment:
Very small nit but I personally love the golden path in these scenarios:

```javascript
if (unprocessedMilestones.length === 0) {
  console.log("No new Gitlab milestones detected");
  return
}

// Actually process milestones here
```

Reply:
Done